File: crates/awsim-conformance/src/runner/route53.rs
use crate::chk;
use crate::runner::common::*;

pub async fn test_route53(endpoint: &str, verbose: bool) -> Vec<OpResult> {
    let config = make_config(endpoint).await;
    let client = aws_sdk_route53::Client::new(&config);
    let mut results = Vec::new();

    // CreateHostedZone
    let create_zone_r = client
        .create_hosted_zone()
        .name("conformance.example.com.")
        .caller_reference(uuid::Uuid::new_v4().to_string())
        .send()
        .await;
    let zone_id = create_zone_r
        .as_ref()
        .ok()
        .and_then(|r| r.hosted_zone.as_ref())
        .map(|z| z.id.clone());
    results.push(chk!("CreateHostedZone", create_zone_r, verbose));

    // ListHostedZones
    results.push(chk!(
        "ListHostedZones",
        client.list_hosted_zones().send().await,
        verbose
    ));

    // GetHostedZoneCount
    results.push(chk!(
        "GetHostedZoneCount",
        client.get_hosted_zone_count().send().await,
        verbose
    ));

    if let Some(ref zid) = zone_id {
        // GetHostedZone
        results.push(chk!(
            "GetHostedZone",
            client.get_hosted_zone().id(zid).send().await,
            verbose
        ));

        // ChangeResourceRecordSets (add an A record)
        results.push(chk!(
            "ChangeResourceRecordSets",
            client
                .change_resource_record_sets()
                .hosted_zone_id(zid)
                .change_batch(
                    aws_sdk_route53::types::ChangeBatch::builder()
                        .changes(
                            aws_sdk_route53::types::Change::builder()
                                .action(aws_sdk_route53::types::ChangeAction::Create)
                                .resource_record_set(
                                    aws_sdk_route53::types::ResourceRecordSet::builder()
                                        .name("www.conformance.example.com.")
                                        .r#type(aws_sdk_route53::types::RrType::A)
                                        .ttl(300)
                                        .resource_records(
                                            aws_sdk_route53::types::ResourceRecord::builder()
                                                .value("1.2.3.4")
                                                .build()
                                                .unwrap(),
                                        )
                                        .build()
                                        .unwrap(),
                                )
                                .build()
                                .unwrap(),
                        )
                        .build()
                        .unwrap(),
                )
                .send()
                .await,
            verbose
        ));

        // ListResourceRecordSets
        results.push(chk!(
            "ListResourceRecordSets",
            client
                .list_resource_record_sets()
                .hosted_zone_id(zid)
                .send()
                .await,
            verbose
        ));

        // DeleteHostedZone
        results.push(chk!(
            "DeleteHostedZone",
            client.delete_hosted_zone().id(zid).send().await,
            verbose
        ));
    } else {
        for op in &[
            "GetHostedZone",
            "ChangeResourceRecordSets",
            "ListResourceRecordSets",
            "DeleteHostedZone",
        ] {
            results.push(OpResult::Skipped(op.to_string()));
        }
    }

    // CreateHealthCheck
    let create_hc_r = client
        .create_health_check()
        .caller_reference(uuid::Uuid::new_v4().to_string())
        .health_check_config(
            aws_sdk_route53::types::HealthCheckConfig::builder()
                .ip_address("1.2.3.4")
                .port(80)
                .r#type(aws_sdk_route53::types::HealthCheckType::Http)
                .resource_path("/")
                .request_interval(30)
                .failure_threshold(3)
                .build()
                .unwrap(),
        )
        .send()
        .await;
    let health_check_id = create_hc_r
        .as_ref()
        .ok()
        .and_then(|r| r.health_check.as_ref())
        .map(|h| h.id.clone());
    results.push(chk!("CreateHealthCheck", create_hc_r, verbose));

    // ListHealthChecks
    results.push(chk!(
        "ListHealthChecks",
        client.list_health_checks().send().await,
        verbose
    ));

    // GetHealthCheck
    if let Some(ref hcid) = health_check_id {
        results.push(chk!(
            "GetHealthCheck",
            client.get_health_check().health_check_id(hcid).send().await,
            verbose
        ));
    } else {
        results.push(OpResult::Skipped("GetHealthCheck".to_string()));
    }

    // ListGeoLocations
    results.push(chk!(
        "ListGeoLocations",
        client.list_geo_locations().send().await,
        verbose
    ));

    // ListReusableDelegationSets
    results.push(chk!(
        "ListReusableDelegationSets",
        client.list_reusable_delegation_sets().send().await,
        verbose
    ));

    // DeleteHealthCheck
    if let Some(ref hcid) = health_check_id {
        results.push(chk!(
            "DeleteHealthCheck",
            client
                .delete_health_check()
                .health_check_id(hcid)
                .send()
                .await,
            verbose
        ));
    } else {
        results.push(OpResult::Skipped("DeleteHealthCheck".to_string()));
    }

    results
}

File: ui/src/lib/api/route53.ts
/**
 * Route53 API client.
 *
 * Wraps the AWSim Route53 REST/XML API
 * (`/2013-04-01/...`) with strongly typed camel-cased shapes for hosted
 * zones, resource record sets, and health checks.
 */

import { ENDPOINT, amzDate, authHeader, loggedFetch } from "$lib/aws";

const SERVICE = "route53";
const BASE = `${ENDPOINT}/2013-04-01`;

// ---------- Types ----------

export interface HostedZone {
  id: string;
  name: string;
  callerReference: string;
  resourceRecordSetCount: number;
  comment?: string;
  privateZone: boolean;
}

export interface ResourceRecord {
  value: string;
}

export interface ResourceRecordSet {
  name: string;
  type: string;
  ttl: number;
  records: ResourceRecord[];
  setIdentifier?: string;
  aliasTarget?: { dnsName: string; hostedZoneId: string };
}

export interface HealthCheck {
  id: string;
  callerReference: string;
  type: string;
  resourcePath?: string;
  fullyQualifiedDomainName?: string;
  port?: number;
  ipAddress?: string;
}

// ---------- Internals ----------

function headers(): Record<string, string> {
  return {
    Authorization: authHeader(SERVICE),
    "X-Amz-Date": amzDate(),
    "Content-Type": "application/xml",
  };
}

async function request(
  method: string,
  path: string,
  action: string,
  body?: string,
): Promise<string> {
  const res = await loggedFetch(SERVICE, action, method, `${BASE}${path}`, {
    method,
    headers: headers(),
    body,
  });
  const text = await res.text();
  if (!res.ok) {
    throw new Error(`Route53 ${action} failed (HTTP ${res.status}): ${text}`);
  }
  return text;
}

function tagValue(xml: string, tag: string): string {
  const match = xml.match(new RegExp(`<${tag}>([^<]*)</${tag}>`));
  return match ? match[1] : "";
}

function tagBlocks(xml: string, tag: string): string[] {
  const out: string[] = [];
  const regex = new RegExp(`<${tag}>([\\s\\S]*?)</${tag}>`, "g");
  let m;
  while ((m = regex.exec(xml)) !== null) out.push(m[1]);
  return out;
}

function escapeXml(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

// ---------- Hosted zones ----------

export async function listHostedZones(): Promise<HostedZone[]> {
  const xml = await request("GET", "/hostedzone", "ListHostedZones");
  return tagBlocks(xml, "HostedZone").map((block) => ({
    id: tagValue(block, "Id"),
    name: tagValue(block, "Name"),
    callerReference: tagValue(block, "CallerReference"),
    resourceRecordSetCount:
      Number(tagValue(block, "ResourceRecordSetCount")) || 0,
    comment: tagValue(block, "Comment") || undefined,
    privateZone: tagValue(block, "PrivateZone") === "true",
  }));
}

export interface HostedZoneDetail extends HostedZone {
  nameServers: string[];
}

export async function getHostedZone(id: string): Promise<HostedZoneDetail> {
  const cleanId = id.replace(/^\/?hostedzone\//, "");
  const xml = await request("GET", `/hostedzone/${cleanId}`, "GetHostedZone");
  const zoneBlock = xml.match(/<HostedZone>([\s\S]*?)<\/HostedZone>/);
  const block = zoneBlock ? zoneBlock[1] : xml;
  const nsBlock = xml.match(/<NameServers>([\s\S]*?)<\/NameServers>/);
  return {
    id: tagValue(block, "Id"),
    name: tagValue(block, "Name"),
    callerReference: tagValue(block, "CallerReference"),
    resourceRecordSetCount:
      Number(tagValue(block, "ResourceRecordSetCount")) || 0,
    comment: tagValue(block, "Comment") || undefined,
    privateZone: tagValue(block, "PrivateZone") === "true",
    nameServers: nsBlock
      ? tagBlocks(nsBlock[1], "NameServer").map((s) => s.trim())
      : [],
  };
}

export interface CreateHostedZoneInput {
  name: string;
  comment?: string;
  privateZone?: boolean;
}

export async function createHostedZone(
  input: CreateHostedZoneInput,
): Promise<{ id: string }> {
  const ref = `cli-${Date.now()}`;
  const body = `<?xml version="1.0" encoding="UTF-8"?>
<CreateHostedZoneRequest xmlns="https://route53.amazonaws.com/doc/2013-04-01/">
  <Name>${escapeXml(input.name)}</Name>
  <CallerReference>${ref}</CallerReference>
  <HostedZoneConfig>
    <Comment>${escapeXml(input.comment ?? "")}</Comment>
    <PrivateZone>${input.privateZone ? "true" : "false"}</PrivateZone>
  </HostedZoneConfig>
</CreateHostedZoneRequest>`;
  const xml = await request("POST", "/hostedzone", "CreateHostedZone", body);
  return { id: tagValue(xml, "Id") };
}

export async function deleteHostedZone(id: string): Promise<void> {
  const cleanId = id.replace(/^\/?hostedzone\//, "");
  await request("DELETE", `/hostedzone/${cleanId}`, "DeleteHostedZone");
}

// ---------- Resource record sets ----------

export async function listResourceRecordSets(
  hostedZoneId: string,
): Promise<ResourceRecordSet[]> {
  const cleanId = hostedZoneId.replace(/^\/?hostedzone\//, "");
  const xml = await request(
    "GET",
    `/hostedzone/${cleanId}/rrset`,
    "ListResourceRecordSets",
  );
  return tagBlocks(xml, "ResourceRecordSet").map((block) => {
    const records: ResourceRecord[] = [];
    const rrBlock = block.match(
      /<ResourceRecords>([\s\S]*?)<\/ResourceRecords>/,
    );
    if (rrBlock) {
      tagBlocks(rrBlock[1], "Value").forEach((v) => records.push({ value: v }));
    }
    const aliasBlock = block.match(/<AliasTarget>([\s\S]*?)<\/AliasTarget>/);
    return {
      name: tagValue(block, "Name"),
      type: tagValue(block, "Type"),
      ttl: Number(tagValue(block, "TTL")) || 0,
      records,
      setIdentifier: tagValue(block, "SetIdentifier") || undefined,
      aliasTarget: aliasBlock
        ? {
            dnsName: tagValue(aliasBlock[1], "DNSName"),
            hostedZoneId: tagValue(aliasBlock[1], "HostedZoneId"),
          }
        : undefined,
    };
  });
}

export interface ChangeRecordInput {
  action: "CREATE" | "DELETE" | "UPSERT";
  name: string;
  type: string;
  ttl: number;
  values: string[];
}

export async function changeResourceRecordSets(
  hostedZoneId: string,
  changes: ChangeRecordInput[],
): Promise<void> {
  const cleanId = hostedZoneId.replace(/^\/?hostedzone\//, "");
  const changeXml = changes
    .map((c) => {
      const valuesXml = c.values
        .map(
          (v) =>
            `<ResourceRecord><Value>${escapeXml(v)}</Value></ResourceRecord>`,
        )
        .join("");
      return `<Change>
        <Action>${c.action}</Action>
        <ResourceRecordSet>
          <Name>${escapeXml(c.name)}</Name>
          <Type>${c.type}</Type>
          <TTL>${c.ttl}</TTL>
          <ResourceRecords>${valuesXml}</ResourceRecords>
        </ResourceRecordSet>
      </Change>`;
    })
    .join("");
  const body = `<?xml version="1.0" encoding="UTF-8"?>
<ChangeResourceRecordSetsRequest xmlns="https://route53.amazonaws.com/doc/2013-04-01/">
  <ChangeBatch><Changes>${changeXml}</Changes></ChangeBatch>
</ChangeResourceRecordSetsRequest>`;
  await request(
    "POST",
    `/hostedzone/${cleanId}/rrset`,
    "ChangeResourceRecordSets",
    body,
  );
}

// ---------- Health checks ----------

export async function listHealthChecks(): Promise<HealthCheck[]> {
  const xml = await request("GET", "/healthcheck", "ListHealthChecks");
  return tagBlocks(xml, "HealthCheck").map((block) => {
    const cfgBlock = block.match(
      /<HealthCheckConfig>([\s\S]*?)<\/HealthCheckConfig>/,
    );
    const cfg = cfgBlock ? cfgBlock[1] : "";
    return {
      id: tagValue(block, "Id"),
      callerReference: tagValue(block, "CallerReference"),
      type: tagValue(cfg, "Type"),
      resourcePath: tagValue(cfg, "ResourcePath") || undefined,
      fullyQualifiedDomainName:
        tagValue(cfg, "FullyQualifiedDomainName") || undefined,
      port: Number(tagValue(cfg, "Port")) || undefined,
      ipAddress: tagValue(cfg, "IPAddress") || undefined,
    };
  });
}

export async function getHealthCheck(id: string): Promise<HealthCheck> {
  const xml = await request("GET", `/healthcheck/${id}`, "GetHealthCheck");
  const cfgBlock = xml.match(
    /<HealthCheckConfig>([\s\S]*?)<\/HealthCheckConfig>/,
  );
  const cfg = cfgBlock ? cfgBlock[1] : "";
  return {
    id: tagValue(xml, "Id"),
    callerReference: tagValue(xml, "CallerReference"),
    type: tagValue(cfg, "Type"),
    resourcePath: tagValue(cfg, "ResourcePath") || undefined,
    fullyQualifiedDomainName:
      tagValue(cfg, "FullyQualifiedDomainName") || undefined,
    port: Number(tagValue(cfg, "Port")) || undefined,
    ipAddress: tagValue(cfg, "IPAddress") || undefined,
  };
}
